authorization/role/service/role_management_service.go

```diff
@@ -50,84 +50,92 @@ func (r *RoleManagementServiceImpl) ListAvailableRolesByResourceType(ctx context
-		assignedRoles, err := r.repo.IdentityRoleRepository().FindIdentityRolesByIdentityAndResource(ctx, resourceID, identityIDAsUUID)
+	assignments := make(map[uuid.UUID][]uuid.UUID)
+
+	for k, v := range roleAssignments {
+		roleRef, err := r.repo.RoleRepository().Lookup(ctx, k, rt.ResourceType.Name)
 		if err != nil {
-			log.Error(ctx, map[string]interface{}{
-				"resource_id": resourceID,
-				"identity_id": assignedBy,
-				"role":        roleName,
-			}, "error looking up existing assignments")
 			return err
 		}
-		if len(assignedRoles) == 0 {
-			log.Error(ctx, map[string]interface{}{
-				"resource_id": resourceID,
-				"identity_id": identityIDAsUUID,
-				"role":        roleName,
-			}, "identity not part of a resource cannot be assigned a role")
-			return errors.NewBadParameterErrorFromString("identityID", identityIDAsUUID, fmt.Sprintf("cannot update roles for an identity %s without an existing role", identityIDAsUUID))
+
+		for _, identityIDAsUUID := range v {
+			assignedRoles, err := r.repo.IdentityRoleRepository().FindIdentityRolesByIdentityAndResource(ctx, resourceID, identityIDAsUUID)
+			if err != nil {
+				log.Error(ctx, map[string]interface{}{
+					"resource_id": resourceID,
+					"identity_id": assignedBy,
+				}, "error looking up existing assignments")
+				return err
+			}
+			if len(assignedRoles) == 0 {
+				log.Error(ctx, map[string]interface{}{
+					"resource_id": resourceID,
+					"identity_id": identityIDAsUUID,
+				}, "identity not previously assigned a resource role cannot be assigned another role")
+				return errors.NewBadParameterErrorFromString("identityID", identityIDAsUUID, fmt.Sprintf("cannot update roles for an identity %s without an existing role", identityIDAsUUID))
+			}
+
+			if ids, found := assignments[roleRef.RoleID]; found {
+				assignments[roleRef.RoleID] = append(ids, identityIDAsUUID)
+			} else {
+				assignments[roleRef.RoleID] = []uuid.UUID{identityIDAsUUID}
+			}
 		}
 	}
 
 	err = transaction.Transactional(r.tm, func(tr transaction.TransactionalResources) error {
 		// Now that we have confirmed that all users have pre-existing role assignments
 		// we can proceed with the assignment of roles.
-		for _, identityIDAsUUID := range identitiesToBeAssigned {
-			ir := rolerepo.IdentityRole{
-				ResourceID: resourceID,
-				IdentityID: identityIDAsUUID,
-				RoleID:     roleRef.RoleID,
-			}
 
-			err = r.repo.IdentityRoleRepository().Create(ctx, &ir)
-
-			if err != nil {
-				log.Error(ctx, map[string]interface{}{
-					"resource_id": resourceID,
-					"identity_id": identityIDAsUUID,
-					"role":        roleName,
-				}, "assignment failed")
-				return err
+		for roleID, ids := range assignments {
+			for _, identityIDAsUUID := range ids {
+				ir := rolerepo.IdentityRole{
+					ResourceID: resourceID,
+					IdentityID: identityIDAsUUID,
+					RoleID:     roleID,
+				}
+
+				err = r.repo.IdentityRoleRepository().Create(ctx, &ir)
```

Comment:
Another thing I missed in the previous reviews. We just append new roles instead of replacing the existing ones.
So, if the user has "Admin" role and we want to replace it by "Viewer" we can't do it via Assign() (and we don't have anything in the service to remove the existing role). If we assign "Viewer" then the user will have both Admin and Viewer roles.

Reply:
Fixed in https://github.com/fabric8-services/fabric8-auth/pull/481/commits/6cb4606c05f00806f5c36f714ff7e3ec371e2e3a